 ## <font color = 'brown'> Data Assignment </font>
<u> Objectives </u>
    <li> Clean and extract valuable information that can be used for dashboarding </li>

In [8]:
#installing plotly to help in quick visulaization
!pip install plotly==4.14.3
!pip install matplotlib

In [14]:
import matplotlib
print(matplotlib.__version__)

3.1.3


In [15]:
#importing the neccesary modules 
import pandas as pd
import numpy as np
from matplotlib import pyplot as plt
import seaborn as sns
import plotly.express as px

In [126]:
# get the dataset
users_df = pd.read_csv("E:/assignment/User_Grid_View.csv")


In [127]:
users_df.head(5)
#printing the size of the data set
size = users_df.shape
print("size of data set is : {}".format(size))

size of data set is : (2093, 19)


In [128]:
#check how manynull values we have
users_df.apply(lambda x: sum(x.isnull()), axis = 0)

ID                         0
Source                     0
Customer Segment         468
Membership Type          373
Membership Level         395
Affiliation              542
IsActive                 380
Gender                   386
Citizenship             2093
Age                     1871
Lifetime Days             13
Has Linkedin            1919
City                     789
Country                  406
Industry                1756
Interviewed             2081
Pathfinder Assigned     1755
Customer Call Status     652
Languages               2019
dtype: int64

*__Observations__*

From the function above we have a lot of null values in the dataset, we are going to omit those we very large values, and they will not have a significant impact on the analysis and remain with the columns that can have features extracted from them 
The columns to be dropped will be:
* Languages
* Citizenship
* Pathfinder Assinged
* Age
* Has LinkedIn
* Industry

We shall sample the data and drop the columns that will not be used in the analysis of the data. 
Based on the data size we shall use *__Simple random sampling technique__*



In [129]:
users_df.head()

ID      Source      Customer Segment   Membership Type Membership Level  \
0   1  RoomMaster  Existing Room member   Founding Member    Senior Leader   
1   2  RoomMaster       LP Nairobi 2018        Free Trial     Young Leader   
2   3  RoomMaster             ALU Staff  Staff Membership    Senior Leader   
3   4  RoomMaster     DSLP Nairobi 2020        Free Trial     Young Leader   
4   5  RoomMaster  Existing Room member        Free Trial     Young Leader   

         Affiliation IsActive  Gender  Citizenship  Age  Lifetime Days  \
0                ALN  checked  female          NaN  NaN          366.0   
1    LP Nairobi 2018  checked  female          NaN  NaN          214.0   
2           AL Staff  checked    male          NaN  NaN          214.0   
3  DSLP Nairobi 2020  checked  female          NaN  NaN          214.0   
4         ALA Alumni  checked  female          NaN  NaN          214.0   

  Has Linkedin          City                   Country   Industry Interviewed  \
0          NaN           NaN  United States of America        NaN         NaN   
1          NaN       Nairobi                     Kenya        NaN         NaN   
2          NaN           NaN                 Mauritius  Education         NaN   
3          NaN           NaN                     Kenya        NaN         NaN   
4          NaN  Johannesburg              South Africa        NaN         NaN   

  Pathfinder Assigned Customer Call Status Languages  
0                 NaN                 None       NaN  
1                 NaN            Completed       NaN  
2                 NaN            Completed       NaN  
3                 NaN                 None       NaN  
4                 NaN                 None       NaN

In [146]:
# we shall use 1800 records for our analysis
users_df_sam = users_df.sample(1800)
users_df_sam

ID      Source      Customer Segment   Membership Type  \
1535  1536  RoomMaster     DSXL Johannesburg        Free Trial   
1195  1196  RoomMaster             AIIP 2019        Free Trial   
619    620  RoomMaster  Existing Room member   Founding Member   
231    232  RoomMaster             ALU staff  Staff Membership   
62      63  RoomMaster  Existing Room member   Founding Member   
...    ...         ...                   ...               ...   
1997  1998     members                   NaN               NaN   
1011  1012  RoomMaster             ALU staff  Staff Membership   
521    522  RoomMaster  Existing Room member        Free Trial   
625    626  RoomMaster  Existing Room member        Free Trial   
1143  1144  RoomMaster                   NaN    Premium Paying   

       Membership Level        Affiliation IsActive  Gender  Citizenship  Age  \
1535  Mid-career Leader  DSXL Johannesburg  checked    male          NaN  NaN   
1195       Young Leader          AIIP 2019  checked    male          NaN  NaN   
619   Mid-career Leader         ALA Alumni  checked    male          NaN  NaN   
231   Mid-career Leader           AL Staff  checked    male          NaN  NaN   
62        Senior Leader                NaN  checked    male          NaN  NaN   
...                 ...                ...      ...     ...          ...  ...   
1997                NaN                NaN      NaN     NaN          NaN  NaN   
1011  Mid-career Leader           AL Staff  checked    male          NaN  NaN   
521       Senior Leader                ALN  checked    male          NaN  NaN   
625       Senior Leader                ALN  checked    male          NaN  NaN   
1143       Young Leader       ALU 2016 ISA  checked  female          NaN  NaN   

      Lifetime Days Has Linkedin          City                   Country  \
1535          214.0          NaN  Johannesburg              South Africa   
1195          214.0          NaN     Lephalale              South Africa   
619           366.0          NaN           NaN              South Africa   
231           214.0          NaN        Kigali                    Rwanda   
62            366.0          NaN        London            United Kingdom   
...             ...          ...           ...                       ...   
1997          119.0          NaN           NaN                       NaN   
1011          214.0          NaN          Mesa  United States of America   
521           214.0          NaN       Nairobi                     Kenya   
625           214.0          NaN        Zurich               Switzerland   
1143           61.0          NaN         Lagos                   Nigeria   

       Industry Interviewed Pathfinder Assigned Customer Call Status Languages  
1535        NaN         NaN                 NaN            Completed       NaN  
1195        NaN         NaN                 NaN            Completed       NaN  
619         NaN         NaN             checked                  NaN       NaN  
231   Education         NaN                 NaN                 None       NaN  
62          NaN         NaN                 NaN                 None       NaN  
...         ...         ...                 ...                  ...       ...  
1997        NaN         NaN                 NaN                  NaN       NaN  
1011  Education         NaN                 NaN            Completed       NaN  
521         NaN         NaN                 NaN                 None       NaN  
625         NaN         NaN                 NaN                 None       NaN  
1143        NaN         NaN             checked                  NaN       NaN  

[1800 rows x 19 columns]

In [147]:
#looking at the null values
print(users_df_sam.apply(lambda x: sum(x.isnull())))

users_df_sam.columns.to_list()

ID                         0
Source                     0
Customer Segment         396
Membership Type          316
Membership Level         337
Affiliation              464
IsActive                 321
Gender                   329
Citizenship             1800
Age                     1606
Lifetime Days              9
Has Linkedin            1652
City                     668
Country                  339
Industry                1504
Interviewed             1789
Pathfinder Assigned     1504
Customer Call Status     551
Languages               1734
dtype: int64


['ID',
 'Source',
 'Customer Segment',
 'Membership Type',
 'Membership Level',
 'Affiliation',
 'IsActive',
 'Gender',
 'Citizenship',
 'Age',
 'Lifetime Days',
 'Has Linkedin',
 'City',
 'Country',
 'Industry',
 'Interviewed',
 'Pathfinder Assigned',
 'Customer Call Status',
 'Languages']

In [148]:
#dropping the columns that will not be useful in the analysis
users_df_sam.drop(['Languages','Customer Call Status','Pathfinder Assigned', 'Interviewed','Industry'
                   ,'City','Has Linkedin','Age','Citizenship'], axis = 1, inplace=True)


In [149]:
# dealing with the missing values in the lifetime Days columns
# we shall use the mean lifetime days to get rid of the null values

users_df_sam["Lifetime Days"].fillna((users_df_sam['Lifetime Days'].mean()), inplace = True)
users_df_sam['Lifetime Days'].isnull().sum()

In [150]:
print(users_df_sam.apply(lambda x: sum(x.isnull())))

ID                    0
Source                0
Customer Segment    396
Membership Type     316
Membership Level    337
Affiliation         464
IsActive            321
Gender              329
Lifetime Days         0
Country             339
dtype: int64


In [154]:
# from this size of 1800 we shall ommit the NaN 
users_df_sam.dropna(inplace=True)

In [155]:
users_df_sam.shape

(1174, 10)

For our analysis we shall use the 1174 sample size, which contains no  missing values

In [167]:
users_df_sam.groupby('Affiliation')['ID'].count()

Affiliation
AIIP 2018                           17
AIIP 2019                           31
AIIP 2020                           27
AL Staff                           143
ALA Alumni                          34
ALN                                269
ALU Alumni                          62
ALUSB 2016 Alumni ISA                7
ALUSB 2017 Alumni ISA                5
ALUSB Alumni                        56
Advisory Council                     8
Board Member                         1
Consultant                           4
Corporate Partner                   18
DSLP Nairobi 2019                    3
DSLP Nairobi 2020                   19
DSXL Johannesburg                   12
Donor                               25
Entrepreneur in Residence            5
Investor                             1
LP Nairobi 2018                     78
LP Nairobi 2019                     47
LP Nairobi 2020                      1
Member Nomination                    1
Parent                              26
School of Wil

In [178]:
users_df_sam.head()

ID      Source      Customer Segment   Membership Type  \
1535  1536  RoomMaster     DSXL Johannesburg        Free Trial   
1195  1196  RoomMaster             AIIP 2019        Free Trial   
619    620  RoomMaster  Existing Room member   Founding Member   
231    232  RoomMaster             ALU staff  Staff Membership   
1365  1366  RoomMaster  Existing Room member        Free Trial   

       Membership Level        Affiliation IsActive Gender  Lifetime Days  \
1535  Mid-career Leader  DSXL Johannesburg  checked   male          214.0   
1195       Young Leader          AIIP 2019  checked   male          214.0   
619   Mid-career Leader         ALA Alumni  checked   male          366.0   
231   Mid-career Leader           AL Staff  checked   male          214.0   
1365  Mid-career Leader       ALUSB Alumni  checked   male          214.0   

           Country Yearly_count  
1535  South Africa               
1195  South Africa               
619   South Africa               
231         Rwanda               
1365         Kenya

In [177]:
# adding a new column inorder to have more complteness for the present state of the values
# we shall use the life time days to assign values in years 
users_df_sam['Yearly_count'] = " "

In [234]:
# creating an empty list and dividing the number of days by 365 days
# to get the number in years, this will help in further analysis of the data
a = []
for i in users_df_sam['Lifetime Days']:
    i  = i /365
    i = round(i,1)
    a.append(i)  
 

In [235]:
users_df_sam['Yearly_count'] = a

In [236]:

users_df_sam.head()

ID      Source      Customer Segment   Membership Type  \
1535  1536  RoomMaster     DSXL Johannesburg        Free Trial   
1195  1196  RoomMaster             AIIP 2019        Free Trial   
619    620  RoomMaster  Existing Room member   Founding Member   
231    232  RoomMaster             ALU staff  Staff Membership   
1365  1366  RoomMaster  Existing Room member        Free Trial   

       Membership Level        Affiliation IsActive Gender  Lifetime Days  \
1535  Mid-career Leader  DSXL Johannesburg  checked   male          214.0   
1195       Young Leader          AIIP 2019  checked   male          214.0   
619   Mid-career Leader         ALA Alumni  checked   male          366.0   
231   Mid-career Leader           AL Staff  checked   male          214.0   
1365  Mid-career Leader       ALUSB Alumni  checked   male          214.0   

           Country  Yearly_count  
1535  South Africa           0.6  
1195  South Africa           0.6  
619   South Africa           1.0  
231         Rwanda           0.6  
1365         Kenya           0.6

### Exploratory Analysis

Looking through the data to extract useful information 

In [254]:
fig = px.histogram(users_df_sam['Yearly_count'], color=users_df_sam['Membership Type'],
                  nbins = 10, hover_name=users_df_sam['Country'], marginal='rug',
                  height=800, width= 1000, title = "Distribution of members based on the years")
fig.update_xaxes(
ticksuffix = "yrs", showgrid = True)
fig.update_layout ( # customize the font and the legend
 font_family = "Rockwell", 
legend = dict (
title = "Sample distribution", orientation = "h"))
fig.show()

### Observations
Quick take aways from the historgram above is that we have a majority of the members are within 6 months range and they are of free trial (621). The members are mostly from Kenya, USA and Switzerland, and premium paids lie within this region but only account for 149 members

__The data shown here excludes the ones from beehive and members who have alot of missing data__

In [255]:
# Exporting the sample data frame

users_df_sam.to_csv(r'E:/assignment/Exported_Data.csv')